zio-http/src/main/scala/zio/http/Http.scala

```diff
@@ -552,87 +461,49 @@ sealed trait Http[-R, +E, -A, +B] { self =>
-   *
-   * NOTE: `execute` is not a stack-safe method for performance reasons. Unlike
-   * ZIO, there is no reason why the execute should be stack safe. The
-   * performance improves quite significantly if no additional heap allocations
-   * are required this way.
-   */
-  final private[zio] def execute(a: A)(implicit trace: Trace): HExit[R, E, B] =
-    self match {
-
-      case Http.Empty                              => HExit.empty
-      case Http.Identity                           => HExit.succeed(a.asInstanceOf[B])
-      case Succeed(b)                              => HExit.succeed(b)
-      case Fail(cause)                             => HExit.failCause(cause)
-      case Attempt(a)                              =>
-        try { HExit.succeed(a()) }
-        catch { case e: Throwable => HExit.fail(e.asInstanceOf[E]) }
-      case FromFunctionHExit(f)                    =>
-        try { f(a) }
-        catch { case e: Throwable => HExit.die(e) }
-      case FromHExit(h)                            => h
-      case Chain(self, other)                      => self.execute(a).flatMap(b => other.execute(b))
-      case Race(self, other)                       =>
-        (self.execute(a), other.execute(a)) match {
-          case (HExit.Effect(self), HExit.Effect(other)) =>
-            Http.fromOptionFunction[Any](_ => self.raceFirst(other)).execute(a)
-          case (HExit.Effect(_), other)                  => other
-          case (self, _)                                 => self
-        }
-      case FoldHttp(self, failure, success, empty) =>
-        try {
-          self.execute(a).foldExit(failure(_).execute(a), success(_).execute(a), empty.execute(a))
-        } catch {
-          case e: Throwable => HExit.die(e)
-        }
+  final private[zio] def executeToZIO(a: A): ZIO[R, E, B] =
+    toHExit(a).toZIO
 
-      case RunMiddleware(app, mid) =>
-        try {
-          mid(app).execute(a)
-        } catch {
-          case e: Throwable => HExit.die(e)
-        }
+  final private[zio] def toHExit(a: A): HExit[R, E, B] = {
+    val hExit = toHExitOrNull(a)
+    if (hExit ne null) hExit
+    else HExit.die(new MatchError(a))
+  }
 
-      case When(f, other) =>
-        try {
-          if (f(a)) other.execute(a) else HExit.empty
-        } catch {
-          case e: Throwable => HExit.die(e)
-        }
+  private[zio] def toHExitOrNull(a: A): HExit[R, E, B] =
+    execute.applyOrElse(a, (_: A) => null)
 
-      case Combine(self, other) => {
-        self.execute(a) match {
-          case HExit.Empty            => other.execute(a)
-          case exit: HExit.Success[_] => exit.asInstanceOf[HExit[R, E, B]]
-          case exit: HExit.Failure[_] => exit.asInstanceOf[HExit[R, E, B]]
-          case exit @ HExit.Effect(_) => exit.defaultWith(other.execute(a)).asInstanceOf[HExit[R, E, B]]
-        }
-      }
-    }
+  private[zio] def execute: PartialFunction[A, HExit[R, E, B]]
```

Comment:
This will necessitate a `PartialFunction` allocation every time `execute` is invoked. The overhead can be eliminated by `extend`ing `PartialFunction`.

Reply:
Made this change; I had to rename the existing `apply` that was returning a `ZIO` to `toZIO`.